contracts/MultiPartyEscrow.sol

```diff
@@ -0,0 +1,270 @@
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+contract MultiPartyEscrow {
+    
+    //it seems we don't need SafeMath
+   //using SafeMath for uint256;
+    
+
+    //TODO: we could use uint64 for replica_id and nonce (it could be cheaper to store but more expensive to operate with)
+
+    //the full ID of "atomic" payment channel = "[this, channel_id, nonce]"
+    struct PaymentChannel {
+        address sender;      // The account sending payments.
+        address recipient;    // The account receiving the payments.
+        uint256 replica_id;  // id of particular service replica
+        uint256 value;       // Total amount of tokens deposited to the channel. 
+        uint256 nonce;       // "nonce" of the channel (by changing nonce we effectivly close the old channel ([this, channel_id, old_nonce])
+                             //  and open the new channel [this, channel_id, new_nonce])
+                             //!!! nonce also prevents race conditon between channel_claim and channel_extend_and_add_funds 
+        uint256 expiration;  // Timeout in case the recipient never closes.
+    }
+
+
+    mapping (uint256 => PaymentChannel) public channels;
+    mapping (address => uint256)        public balances; //tokens which have been deposit but haven't been escrowed in the channels
+   
+    //already used messages for open_channel_by_recipient in order to prevent replay attack
+    mapping (bytes32 => bool) public used_messages; 
```

Comment:
This is the summary of what we have discussed with Sergey:
- to prevent ```open_channel_by_recipient()``` replay attack contract should store all of messages which were already used by server; if operation is used frequently this store grows without restriction
- if ```open_channel_by_recipient``` is rarely used then it is not clear why client cannot open channel itself
- coding complexity on client side can be managed by SDK
- ideally snet-daemon should not do blockchain operations frequently because blockchain is slow and operation time is not guaranteed. 

Taking this into account, I would propose don't implement open_channel_by_recipient in first drop. It should be carefully designed before implementing it.

Reply:
Given the above, I agree with @vsbogd on postponing this until a careful design solution is found that doesn't involve loads of storage or other overhead.

Reply:
We've discussed with Vitaliy how we could design it for different scenario. I will described it in separate Issue.

Reply:
@astroseger , I would propose removing ```open_channel_by_recipient()``` related code and add it again when we have final design for the issue https://github.com/singnet/platform-contracts/issues/42 .